File: app/src/main/java/catchla/yep/service/MessageService.java
package catchla.yep.service;

import android.accounts.Account;
import android.app.Service;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Intent;
import android.os.IBinder;
import android.util.Log;

import com.desmond.asyncmanager.AsyncManager;
import com.desmond.asyncmanager.PersistedTaskRunnable;
import com.desmond.asyncmanager.TaskRunnable;
import com.squareup.otto.Bus;

import java.util.ArrayList;
import java.util.HashMap;

import catchla.yep.BuildConfig;
import catchla.yep.Constants;
import catchla.yep.message.MessageRefreshedEvent;
import catchla.yep.model.Circle;
import catchla.yep.model.Conversation;
import catchla.yep.model.Friendship;
import catchla.yep.model.Message;
import catchla.yep.model.PagedFriendships;
import catchla.yep.model.PagedMessages;
import catchla.yep.model.Paging;
import catchla.yep.model.TaskResponse;
import catchla.yep.model.User;
import catchla.yep.provider.YepDataStore.Conversations;
import catchla.yep.provider.YepDataStore.Friendships;
import catchla.yep.provider.YepDataStore.Messages;
import catchla.yep.util.ContentResolverUtils;
import catchla.yep.util.ContentValuesCreator;
import catchla.yep.util.JsonSerializer;
import catchla.yep.util.Utils;
import catchla.yep.util.YepAPI;
import catchla.yep.util.YepAPIFactory;
import catchla.yep.util.YepException;

/**
 * Created by mariotaku on 15/5/29.
 */
public class MessageService extends Service implements Constants {

    public static final String ACTION_PREFIX = BuildConfig.APPLICATION_ID + ".";
    public static final String ACTION_REFRESH_MESSAGES = ACTION_PREFIX + "REFRESH_MESSAGES";
    public static final String ACTION_REFRESH_FRIENDSHIPS = ACTION_PREFIX + "REFRESH_FRIENDSHIPS";

    @Override
    public IBinder onBind(final Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(final Intent intent, final int flags, final int startId) {
        if (intent == null) return START_NOT_STICKY;
        final String action = intent.getAction();
        if (action == null) return START_NOT_STICKY;
        switch (action) {
            case ACTION_REFRESH_FRIENDSHIPS: {
                refreshFriendships();
                break;
            }
            case ACTION_REFRESH_MESSAGES: {
                refreshMessages();
                break;
            }
        }
        return START_STICKY;
    }

    private void refreshFriendships() {
        final Account account = Utils.getCurrentAccount(this);
        if (account == null) return;
        final TaskRunnable<Account, TaskResponse<Boolean>, MessageService>
                task = new PersistedTaskRunnable<Account, TaskResponse<Boolean>, MessageService>() {
            @Override
            public TaskResponse<Boolean> doLongOperation(final Account account) throws InterruptedException {
                final YepAPI yep = YepAPIFactory.getInstance(getApplication(), account);
                try {
                    PagedFriendships friendships;
                    int page = 1;
                    final Paging paging = new Paging();
                    final ArrayList<ContentValues> values = new ArrayList<>();
                    while ((friendships = yep.getFriendships(paging)).size() > 0) {
                        for (Friendship friendship : friendships) {
                            values.add(ContentValuesCreator.fromFriendship(friendship));
                        }
                        paging.page(++page);
                        if (friendships.getCount() < friendships.getPerPage()) break;
                    }
                    final ContentResolver cr = getContentResolver();
                    cr.delete(Friendships.CONTENT_URI, null, null);
                    ContentResolverUtils.bulkInsert(cr, Friendships.CONTENT_URI, values);
                    return TaskResponse.getInstance(true);
                } catch (YepException e) {
                    Log.w(LOGTAG, e);
                    return TaskResponse.getInstance(e);
                } finally {
                }
            }

            @Override
            public void callback(final TaskResponse<Boolean> response) {
                final Bus bus = Utils.getMessageBus();
                bus.post(new MessageRefreshedEvent());
            }

            @Override
            public void callback(final MessageService messageService, final TaskResponse<Boolean> response) {
                callback(response);
            }
        };
        task.setParams(account);
        task.setResultHandler(this);
        AsyncManager.runBackgroundTask(task);
    }

    private void refreshMessages() {
        final Account account = Utils.getCurrentAccount(this);
        if (account == null) return;
        final TaskRunnable<Account, TaskResponse<Boolean>, MessageService>
                task = new PersistedTaskRunnable<Account, TaskResponse<Boolean>, MessageService>() {
            @Override
            public TaskResponse<Boolean> doLongOperation(final Account account) throws InterruptedException {
                final YepAPI yep = YepAPIFactory.getInstance(getApplication(), account);
                try {
                    PagedMessages messages;
                    int page = 1;
                    final Paging paging = new Paging();
                    HashMap<String, ContentValues> conversations = new HashMap<>();
                    while ((messages = yep.getUnreadMessages(paging)).size() > 0) {
                        for (Message message : messages) {
                            final String recipientType = message.getRecipientType();
                            final String conversationId = Conversation.generateId(message);
                            message.setConversationId(conversationId);
                            message.setOutgoing(false);

                            ContentValues conversation = conversations.get(conversationId);
                            if (conversation == null) {
                                conversation = new ContentValues();
                                conversation.put(Conversations.CIRCLE, JsonSerializer.serialize(message.getCircle(), Circle.class));
                                conversation.put(Conversations.USER, JsonSerializer.serialize(message.getSender(), User.class));
                                conversation.put(Conversations.RECIPIENT_TYPE, recipientType);
                                conversation.put(Conversations.CONVERSATION_ID, conversationId);
                                conversations.put(conversationId, conversation);
                            }
                            conversation.put(Conversations.UPDATED_AT, message.getCreatedAt().getTime());
                            conversation.put(Conversations.TEXT_CONTENT, message.getTextContent());
                        }

                        paging.page(++page);
                        if (messages.getCount() < messages.getPerPage()) break;
                    }
                    final ArrayList<ContentValues> messagesValues = new ArrayList<>();
                    for (Message message : messages) {
                        messagesValues.add(ContentValuesCreator.fromMessage(message));
                    }
                    final ContentResolver cr = getContentResolver();
                    ContentResolverUtils.bulkInsert(cr, Messages.CONTENT_URI, messagesValues);
                    ContentResolverUtils.bulkInsert(cr, Conversations.CONTENT_URI, conversations.values());
                    return TaskResponse.getInstance(true);
                } catch (YepException e) {
                    Log.w(LOGTAG, e);
                    return TaskResponse.getInstance(e);
                } finally {
                }
            }

            @Override
            public void callback(final TaskResponse<Boolean> response) {
                final Bus bus = Utils.getMessageBus();
                bus.post(new MessageRefreshedEvent());
            }

            @Override
            public void callback(final MessageService messageService, final TaskResponse<Boolean> response) {
                callback(response);
            }
        };
        task.setParams(account);
        task.setResultHandler(this);
        AsyncManager.runBackgroundTask(task);
    }
}

File: src/main/kotlin/catchla/yep/service/MessageService.kt
package catchla.yep.service

import android.accounts.Account
import android.app.Service
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.support.v4.util.SimpleArrayMap
import android.text.TextUtils
import catchla.yep.BuildConfig
import catchla.yep.Constants
import catchla.yep.message.FriendshipsRefreshedEvent
import catchla.yep.message.MessageRefreshedEvent
import catchla.yep.model.*
import catchla.yep.provider.YepDataStore.*
import catchla.yep.util.Utils
import catchla.yep.util.YepAPIFactory
import catchla.yep.util.dagger.GeneralComponentHelper
import com.squareup.otto.Bus
import nl.komponents.kovenant.task
import nl.komponents.kovenant.ui.successUi
import org.mariotaku.ktextension.bulkInsertSliced
import org.mariotaku.sqliteqb.library.Expression
import java.util.*
import javax.inject.Inject

/**
 * Created by mariotaku on 15/5/29.
 */
class MessageService : Service(), Constants {

    @Inject
    lateinit var bus: Bus

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        GeneralComponentHelper.build(this).inject(this)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent == null) return Service.START_NOT_STICKY
        val action = intent.action ?: return Service.START_NOT_STICKY
        when (action) {
            ACTION_REFRESH_FRIENDSHIPS -> {
                val account = intent.getParcelableExtra<Account>(Constants.EXTRA_ACCOUNT)
                refreshFriendships(account)
            }
            ACTION_REFRESH_MESSAGES -> {
                refreshCircles()
                refreshMessages()
            }
            ACTION_REFRESH_USER_INFO -> {
                val account = intent.getParcelableExtra<Account>(Constants.EXTRA_ACCOUNT)
                refreshUserInfo(account)
            }
        }
        return Service.START_STICKY
    }

    private fun refreshUserInfo(account: Account?) {
        if (account == null) return

        task {
            with(YepAPIFactory.getInstance(application, account)) {
                return@task getUser()
            }
        }.successUi {
            Utils.saveUserInfo(this, account, it)
        }
    }

    private fun refreshFriendships(account: Account?) {
        if (account == null) return
        task {
            val yep = YepAPIFactory.getInstance(application, account)
            var friendships: ResponseList<Friendship>
            var page = 1
            val paging = Paging()
            val values = ArrayList<ContentValues>()
            val accountId = Utils.getAccountId(application, account)
            do {
                friendships = yep.getFriendships(paging)
                if (friendships.isEmpty()) break
                for (friendship in friendships) {
                    friendship.accountId = accountId
                    values.add(FriendshipValuesCreator.create(friendship))
                }
                paging.page(++page)
            } while (friendships.count >= friendships.perPage)

            val cr = contentResolver
            cr.delete(Friendships.CONTENT_URI, null, null)
            cr.bulkInsertSliced(Friendships.CONTENT_URI, values)
        }.successUi {
            bus.post(FriendshipsRefreshedEvent())
        }
    }

    private fun refreshMessages() {
        val account = Utils.getCurrentAccount(this) ?: return
        val accountUser = Utils.getAccountUser(this, account)
        task {
            val yep = YepAPIFactory.getInstance(application, account)
            val paging = Paging()
            paging.perPage(30)
            val conversations = yep.getConversations(paging)
            insertConversations(this@MessageService, conversations, accountUser.id)
        }.successUi {
            bus.post(MessageRefreshedEvent(account))
        }
    }

    private fun refreshCircles() {
        val account = Utils.getCurrentAccount(this) ?: return
        val accountUser = Utils.getAccountUser(this, account)
        task {
            val yep = YepAPIFactory.getInstance(application, account)
            val paging = Paging()
            val circles = yep.getCircles(paging)
            val accountId = accountUser.id
            insertCircles(this@MessageService, circles, accountId)
        }.successUi {
            bus.post(MessageRefreshedEvent(account))
        }
    }

    companion object {

        @JvmField val ACTION_PREFIX = BuildConfig.APPLICATION_ID + "."
        @JvmField val ACTION_REFRESH_MESSAGES = ACTION_PREFIX + "REFRESH_MESSAGES"
        @JvmField val ACTION_REFRESH_USER_INFO = ACTION_PREFIX + "REFRESH_USER_INFO"
        @JvmField val ACTION_REFRESH_FRIENDSHIPS = ACTION_PREFIX + "REFRESH_FRIENDSHIPS"

        fun insertCircles(context: Context, circles: Collection<Circle>, accountId: String) {
            val cr = context.contentResolver
            cr.delete(Circles.CONTENT_URI, null, null)
            val contentValues = ArrayList<ContentValues>()
            for (circle in circles) {
                circle.accountId = accountId
                contentValues.add(CircleValuesCreator.create(circle))
            }
            cr.bulkInsertSliced(Circles.CONTENT_URI, contentValues)
        }

        fun insertConversations(context: Context, conversations: ConversationsResponse,
                                accountId: String) {
            val conversationsMap = HashMap<String, Conversation>()
            val cr = context.contentResolver
            val messages = conversations.messages
            val users = HashMap<String, User>()
            for (user in conversations.users) {
                users.put(user.id, user)
            }
            val messageValues = arrayOfNulls<ContentValues>(messages.size)
            val randomIds = SimpleArrayMap<String, ContentValues>()
            for (i in 0 until messages.size) {
                val message = messages[i]
                val recipientType = message.recipientType
                val conversationId = Conversation.generateId(message, accountId)
                message.accountId = accountId
                message.conversationId = conversationId

                val values = MessageValuesCreator.create(message)
                messageValues[i] = values
                val randomId = message.randomId
                if (randomId != null) {
                    randomIds.put(randomId, values)
                }

                var conversation: Conversation? = conversationsMap[conversationId]
                val newConversation = conversation == null
                if (conversation == null) {
                    conversation = Conversation.query(cr, accountId, conversationId)
                    if (conversation == null) {
                        conversation = Conversation()
                        conversation.accountId = accountId
                        conversation.id = conversationId
                    }
                }
                val createdAt = message.createdAt
                if (newConversation || greaterThen(createdAt, conversation.updatedAt)) {
                    conversation.textContent = message.textContent
                    val sender = message.sender
                    if (Message.RecipientType.USER == recipientType) {
                        // Outgoing
                        if (!TextUtils.equals(accountId, sender.id)) {
                            conversation.user = sender
                        } else {
                            val user = users[message.recipientId]
                            if (user != null) {
                                conversation.user = user
                            }
                        }
                    } else {
                        conversation.user = sender
                    }
                    conversation.sender = sender
                    conversation.circle = getMessageCircle(context, message, conversations, accountId)
                    conversation.updatedAt = createdAt
                    conversation.recipientType = recipientType
                    conversation.mediaType = message.mediaType
                }
                if (newConversation && conversation.user != null) {
                    conversationsMap.put(conversationId, conversation)
                }
            }

            val conversationValues = ArrayList<ContentValues>()
            for (conversation in conversationsMap.values) {
                conversationValues.add(ConversationValuesCreator.create(conversation))
            }
            cr.bulkInsertSliced(Conversations.CONTENT_URI, conversationValues)

            val updateSentSelection = Expression.and(Expression.equalsArgs(Messages.ACCOUNT_ID),
                    Expression.equalsArgs(Messages.RANDOM_ID)).sql
            val updateSentSelectionArgs = arrayOf<String?>(accountId, null)
            for (i in 0 until randomIds.size()) {
                updateSentSelectionArgs[i] = randomIds.keyAt(i)
                cr.update(Messages.CONTENT_URI, randomIds.valueAt(i), updateSentSelection,
                        updateSentSelectionArgs)
            }
            cr.bulkInsertSliced(Messages.CONTENT_URI, messageValues)
        }

        private fun greaterThen(createdAt: Date?, updatedAt: Date?): Boolean {
            if (updatedAt == null) return createdAt != null
            return createdAt != null && createdAt > updatedAt
        }

        private fun getMessageCircle(context: Context, message: Message,
                                     conversations: ConversationsResponse?,
                                     accountId: String): Circle? {
            if (Message.RecipientType.CIRCLE != message.recipientType) return null
            var circle: Circle? = message.circle
            // First try find in conversations
            if (conversations != null) {
                val circles = conversations.circles
                if (circles != null) {
                    for (item in circles) {
                        if (TextUtils.equals(item.id, message.recipientId)) {
                            return item
                        }
                    }
                }
            }
            // Then try to load from database
            val circleId: String
            if (circle == null) {
                circleId = message.recipientId
            } else if (circle.topic == null) {
                circleId = circle.id
            } else {
                return circle
            }
            val where = Expression.and(Expression.equalsArgs(Circles.ACCOUNT_ID),
                    Expression.equalsArgs(Circles.CIRCLE_ID)).sql
            val whereArgs = arrayOf(accountId, circleId)
            val c = context.contentResolver.query(Circles.CONTENT_URI,
                    CircleTableInfo.COLUMNS, where, whereArgs, null)
            try {
                if (c != null && c.moveToFirst()) {
                    val ci = CircleCursorIndices(c)
                    circle = ci.newObject(c)
                }
            } finally {
                Utils.closeSilently(c)
            }
            return circle
        }
    }
}
